### PR TITLE
Introduce dalek signature scheme

### PR DESCRIPTION
**Pull request by bkolad**
This PR introduces the following changes:
1. `ed25519_dalek` for default signature scheme
2. transactions in tests are now signed with the private key
3. genesis `Config` for `ValueSetter` & `Election` modules
4. `create_config` refactoring (before, we had two functions with similar logic)

### CODE REVIEW

File: demo-app/src/app.rs [hunk 1]
```diff
@@ -23,26 +25,18 @@ use std::path::Path;
...
-
-#[cfg(test)]
 pub const SEQUENCER_DA_ADDRESS: [u8; 32] = [1; 32];
-#[cfg(test)]
 pub const LOCKED_AMOUNT: u64 = 200;
-#[cfg(test)]
 pub const SEQ_PUB_KEY_STR: &str = "seq_pub_key";
-#[cfg(test)]
-pub const TOKEN_NAME: &str = "Token0";
+pub const TOKEN_NAME: &str = "test-token";
```
Comment (citizen-stig) [inline demo-app/src/app.rs:39]: Maybe rename it to SovereignToken, so all airdrop hunters will start using SDK?

File: demo-app/src/app.rs [hunk 2]
```diff
@@ -91,93 +85,77 @@ impl StateTransitionRunner<ZkConfig> for DemoAppRunner<ZkDefaultContext> {
...
+    )
 }
 
-pub fn create_demo_genesis_config<C: Context>(
+#[cfg(feature = "native")]
+fn create_genesis_config<C: Context>(
+    initial_sequencer_balance: u64,
     sequencer_address: C::Address,
     sequencer_da_address: Vec<u8>,
+    value_setter_admin_private_key: &DefaultPrivateKey,
```
Comment (citizen-stig) [inline demo-app/src/app.rs:107]: Why it is called `value_setter_admin_private_key` and not just `admin_private_key`? When I see `setter` I expect some kind of callable, like Fn

File: sov-modules/sov-modules-api/src/default_signature.rs [hunk 3]
```diff
@@ -1,50 +1,100 @@
...
 pub struct DefaultSignature {
-    pub msg_sig: Vec<u8>,
-    pub should_fail: bool,
+    pub msg_sig: DalekSignature,
+}
+
+impl BorshDeserialize for DefaultSignature {
+    fn deserialize_reader<R: std::io::Read>(reader: &mut R) -> std::io::Result<Self> {
+        let mut buffer = [0; SIGNATURE_LENGTH];
+        reader.read_exact(&mut buffer).unwrap();
```
Comment (citizen-stig) [inline sov-modules/sov-modules-api/src/default_signature.rs:73]: Is it intend to panic?
```suggestion
        reader.read_exact(&mut buffer)?;
```

File: sov-modules/sov-modules-api/src/default_signature.rs [hunk 4]
```diff
@@ -1,50 +1,100 @@
...
+
+        Ok(Self {
+            msg_sig: DalekSignature::from_bytes(&buffer).unwrap(),
```
Comment (citizen-stig) [inline sov-modules/sov-modules-api/src/default_signature.rs:76]: Same here about panic:

```suggestion
            msg_sig: DalekSignature::from_bytes(&buffer)?,
```

Comment (citizen-stig) [inline sov-modules/sov-modules-api/src/default_signature.rs:76]: Same here about panic:

```suggestion
            msg_sig: DalekSignature::from_bytes(&buffer)?,
```

File: sov-modules/sov-modules-api/src/default_signature.rs [hunk 5]
```diff
@@ -1,50 +1,100 @@
...
-        Self { pub_key: key }
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct DefaultPublicKey {
+    pub(crate) pub_key: DalekPublicKey,
+}
+
+impl BorshDeserialize for DefaultPublicKey {
+    fn deserialize_reader<R: std::io::Read>(reader: &mut R) -> std::io::Result<Self> {
+        let mut buffer = [0; PUBLIC_KEY_LENGTH];
+        reader.read_exact(&mut buffer).unwrap();
```
Comment (citizen-stig) [inline sov-modules/sov-modules-api/src/default_signature.rs:51]: Is it intend to panic? 

```suggestion
        reader.read_exact(&mut buffer)?;
```

File: sov-modules/sov-modules-api/src/default_signature.rs [hunk 6]
```diff
@@ -1,50 +1,100 @@
...
+
+        Ok(Self {
+            pub_key: DalekPublicKey::from_bytes(&buffer).unwrap(),
```
Comment (citizen-stig) [inline sov-modules/sov-modules-api/src/default_signature.rs:54]: Same here about panic:

```suggestion
            msg_sig: DalekSignature::from_bytes(&buffer)?,
```

File: sov-modules/sov-modules-api/src/default_signature.rs [hunk 7]
```diff
@@ -1,50 +1,100 @@
...
     ) -> Result<(), SigVerificationError> {
-        if self.should_fail {
-            Err(SigVerificationError::BadSignature)
-        } else {
-            Ok(())
-        }
+        pub_key
+            .pub_key
+            .verify_strict(&msg_hash, &self.msg_sig)
+            .map_err(|_| SigVerificationError::BadSignature)
```
Comment (citizen-stig) [inline sov-modules/sov-modules-api/src/default_signature.rs:98]: We swallowing exact error, which is not ideal, but we don't have logging anyway. 
Maybe `BadSignature` at least could have string inside?

File: demo-app/src/app.rs (hunk 2 above)
Comment (citizen-stig) [reply demo-app/src/app.rs:107]: Ok, I see. Can you please add comment on why it is called like that then?

File: sov-modules/sov-modules-api/src/default_signature.rs (hunk 7 above)
Comment (citizen-stig) [reply sov-modules/sov-modules-api/src/default_signature.rs:100]: Sorry, I don't see the change.

**Review by citizen-stig — APPROVED** [review]

File: sov-modules/sov-modules-api/src/default_signature.rs [hunk 8]
```diff
@@ -1,50 +1,102 @@
...
+
+        Ok(Self { msg_sig })
+    }
+}
+
+impl BorshSerialize for DefaultSignature {
+    fn serialize<W: std::io::Write>(&self, writer: &mut W) -> std::io::Result<()> {
+        writer.write_all(self.msg_sig.as_bytes())
+    }
 }
```
Comment (preston-evans98) [inline sov-modules/sov-modules-api/src/default_signature.rs:87]: Same question here - do we have test coverage?

File: sov-modules/sov-modules-api/src/default_signature.rs (hunk 7 above)
Comment (bkolad) [reply sov-modules/sov-modules-api/src/default_signature.rs:100]: It is there but PR diff is not showing it: https://github.com/Sovereign-Labs/sovereign/blob/blaze/dalek_sig/sov-modules/sov-modules-api/src/default_signature.rs#L100

**Pull request merged**